File: mlradar/Chapter04/ovals_of_cassini_example.m
%% Cassini ovals example
% Created by: Lee A. Harrison
% On: 7/1/2018

clear, clc

% Number of points in the curve
N = 1e6;

% Boltzmann's constant
k = 1.38064852e-23;

% Speed of light (ms)
c = 299792458;

% Receiver noise figure (dB)
noise_figure = 6;

% System temperature (K)
system_temperature = 290;

% Separation distance (m)
D = 100;

% Peak transmit power (W)
peak_power = 50.0e3;

% Tx/Rx antenna gain (dB)
transmit_antenna_gain = 20.0;
receive_antenna_gain = 10.0;

% Operating frequency (Hz)
frequency = 1.0e9;

% Bistatic target radar cross section (dBsm)
bistatic_target_rcs = 10.0;

% Receiver bandwidth (Hz)
bandwidth = 10.0e6;

% Tx/Rx losses (dB)
transmit_losses = 4;
receive_losses = 3;

% Calculate the wavelength (m)
wavelength = 3e8 / frequency;

% Parameters for Cassini ovals equation
% r^4 + a^4 - 2 a^2 r^2 (1 + cos(2 theta)) = b^4
a = D * 0.5;

bistatic_range_factor = (peak_power * transmit_antenna_gain * receive_antenna_gain * wavelength^2 * bistatic_target_rcs)/...
    ((4.0 * pi)^3 * k * system_temperature * bandwidth * lin(noise_figure) * transmit_losses * receive_losses);

% Full angle sweep
t = linspace(0, 2*pi, N);

% Calculate the point at which a = b
SNR_0 = db(16 * bistatic_range_factor / D^4);

% The list of desired SNRs
SNR = SNR_0 - [6, 3, 0, -3];

% Convert to linear units
SNR = lin(SNR);

% Create the plot
figure; hold on;

% Loop over all the desired SNR curves
for i = 1:length(SNR)
    
    % Parameter for the Cassini ovals
    b = (bistatic_range_factor / SNR(i))^0.25;
    
    % Special case when a > b
    if a > b
        
        % Caculat the +/- curves
        R1 = sqrt(a^2 * (cos(2*t) + sqrt(cos(2*t).^2 - 1 + (b/a)^4)));
        R2 = sqrt(a^2 * (cos(2*t) - sqrt(cos(2*t).^2 - 1 + (b/a)^4)));
        
        % Indices for imaginary parts == 0
        ir1 = imag(R1) == 0;
        ir2 = imag(R2) == 0;
        
        % Plot both parts of the curve
        plot(R1(ir1).*cos(t(ir1)), R1(ir1).*sin(t(ir1)), 'k.');
        plot(R2(ir2).*cos(t(ir2)), R2(ir2).*sin(t(ir2)), 'k.');        
        
    else
        
        % Calculate the range for continuous curves
        R = sqrt(a^2 * cos(2*t) + sqrt(b^4 - a^4 .* sin(2*t).^2));
        plot(real(R.*cos(t)), real(R.*sin(t)), '.');
        
    end
    
    % Update the text for the legend
    legend_text{i} = sprintf('SNR = %.1f\n', db(SNR(i)));
    
end

% Add the titles and labels.
title('Ovals of Cassini');
xlabel('Range (km)');
ylabel('Range (km)');

% Add the Tx/Rx locations
text(-a, 0, 'Tx');
text( a, 0, 'Rx');

% Turn on the grid
grid on;

% Add the legend
legend(legend_text);

% Use the plot default settings
plot_settings;

File: pyradar/Chapter04/ovals_of_cassini_example.py
"""
Project: RadarBook
File: ovals_of_cassini_example.py
Created by: Lee A. Harrison
On: 7/2/2018
Created with: PyCharm

Copyright (C) 2019 Artech House (artech@artechhouse.com)
This file is part of Introduction to Radar Using Python and MATLAB
and can not be copied and/or distributed without the express permission of Artech House.
"""
import sys
from Chapter04.ui.OvalsOfCassini_ui import Ui_MainWindow
from numpy import linspace, log10, sqrt, sin, cos, imag, real
from scipy.constants import c, Boltzmann as k, pi
from PyQt5.QtWidgets import QApplication, QMainWindow
from matplotlib.backends.qt_compat import QtCore
from matplotlib.backends.backend_qt5agg import (FigureCanvas, NavigationToolbar2QT as NavigationToolbar)
from matplotlib.figure import Figure


class OvalsOfCassini(QMainWindow, Ui_MainWindow):
    def __init__(self):

        super(self.__class__, self).__init__()

        self.setupUi(self)

        # Connect to the input boxes, when the user presses enter the form updates
        self.separation_distance.returnPressed.connect(self._update_canvas)
        self.system_temperature.returnPressed.connect(self._update_canvas)
        self.bandwidth.returnPressed.connect(self._update_canvas)
        self.noise_figure.returnPressed.connect(self._update_canvas)
        self.transmit_losses.returnPressed.connect(self._update_canvas)
        self.receive_losses.returnPressed.connect(self._update_canvas)
        self.peak_power.returnPressed.connect(self._update_canvas)
        self.transmit_antenna_gain.returnPressed.connect(self._update_canvas)
        self.receive_antenna_gain.returnPressed.connect(self._update_canvas)
        self.frequency.returnPressed.connect(self._update_canvas)
        self.bistatic_target_rcs.returnPressed.connect(self._update_canvas)

        # Set up a figure for the plotting canvas
        fig = Figure() 
        self.axes1 = fig.add_subplot(111)
        self.my_canvas = FigureCanvas(fig)

        # Add the canvas to the vertical layout
        self.verticalLayout.addWidget(self.my_canvas)
        self.addToolBar(QtCore.Qt.TopToolBarArea, NavigationToolbar(self.my_canvas, self))

        # Update the canvas for the first display
        self._update_canvas()

    def _update_canvas(self):
        """
        Update the figure when the user changes an input value.
        :return:
        """
        # Get the values from the form
        separation_distance = float(self.separation_distance.text())
        system_temperature = float(self.system_temperature.text())
        bandwidth = float(self.bandwidth.text())
        noise_figure = 10 ** (float(self.noise_figure.text()) / 10)
        transmit_losses = 10 ** (float(self.transmit_losses.text()) / 10)
        receive_losses = 10 ** (float(self.receive_losses.text()) / 10)
        peak_power = float(self.peak_power.text())
        transmit_antenna_gain = 10 ** (float(self.transmit_antenna_gain.text()) / 10)
        receive_antenna_gain = 10 ** (float(self.receive_antenna_gain.text()) / 10)
        frequency = float(self.frequency.text())
        bistatic_target_rcs = 10 ** (float(self.bistatic_target_rcs.text()) / 10)

        # Number of points for plotting ovals
        number_of_points = 100000

        # Parameters for the Cassini ovals equation
        # r ^ 4 + a ^ 4 - 2 a ^ 2 r ^ 2(1 + cos(2 theta)) = b ^ 4

        # Parameter "a"
        a = 0.5 * separation_distance

        # Calculate the wavelength (m)
        wavelength = c / frequency

        # Calculate the bistatic radar range factor
        bistatic_range_factor = (peak_power * transmit_antenna_gain * receive_antenna_gain * wavelength ** 2 *
                                 bistatic_target_rcs) / ((4.0 * pi) ** 3 * k * system_temperature * bandwidth *
                                                         noise_figure * transmit_losses * receive_losses)

        # Full angle sweep
        t = linspace(0, 2.0 * pi, number_of_points)

        # Calculate the signal to noise ratio at which a = b
        SNR_0 = 10.0 * log10(16.0 * bistatic_range_factor / separation_distance ** 4)

        # Create the list of signal to noise ratios to plot
        SNR = [SNR_0 - 6, SNR_0 - 3, SNR_0, SNR_0 + 3]

        # Clear the axes for the updated plot
        self.axes1.clear()

        # Loop over all the desired signal to noise ratios
        for s in SNR:

            # Convert to linear units
            snr = 10.0 ** (s / 10.0)

            # Parameter for Cassini ovals
            b = (bistatic_range_factor / snr) ** 0.25

            if a > b:

                # Calculate the +/- curves
                r1 = sqrt(a ** 2 * (cos(2.0 * t) + sqrt(cos(2 * t) ** 2 - 1.0 + (b / a) ** 4)))
                r2 = sqrt(a ** 2 * (cos(2.0 * t) - sqrt(cos(2 * t) ** 2 - 1.0 + (b / a) ** 4)))

                # Find the correct indices for imaginary parts = 0
                i1 = imag(r1) == 0
                i2 = imag(r2) == 0

                r1 = real(r1)
                r2 = real(r2)

                # Plot both parts of the curve
                label_text = "SNR = {:.1f}".format(s)
                self.axes1.plot(r1[i1] * cos(t[i1]) / 1e3, r1[i1] * sin(t[i1]) / 1e3, 'k.', label=label_text)
                self.axes1.plot(r2[i2] * cos(t[i2]) / 1e3, r2[i2] * sin(t[i2]) / 1e3, 'k.')

            else:

                # Calculate the range for the continuous curves
                r = sqrt(a**2 * cos(2 * t) + sqrt(b ** 4 - a ** 4 * sin(2.0 * t) ** 2))

                # Plot the continuous parts
                label_text = "SNR = {:.1f}".format(s)
                self.axes1.plot(r * cos(t) / 1e3, r * sin(t) / 1e3, '.', label=label_text)

            # Add the text for Tx/Rx locations
            self.axes1.text(-a / 1e3, 0, 'Tx')
            self.axes1.text(a / 1e3, 0, 'Rx')

            # Set the plot title and labels
            self.axes1.set_title('Ovals of Cassini', size=14)
            self.axes1.set_xlabel('Range (km)', size=12)
            self.axes1.set_ylabel('Range (km)', size=12)

            # Set the tick label size
            self.axes1.tick_params(labelsize=12)

            # Turn on the grid
            self.axes1.grid(linestyle=':', linewidth=0.5)

            # Add the legend
            self.axes1.legend(loc='upper left', prop={'size': 10})

            # Update the canvas
            self.my_canvas.draw()


def start():
    form = OvalsOfCassini()  # Set the form
    form.show()              # Show the form


def main():
    app = QApplication(sys.argv)  # A new instance of QApplication
    form = OvalsOfCassini()       # Set the form
    form.show()                   # Show the form
    app.exec_()                   # Execute the app


if __name__ == '__main__':
    main()
